File: backend/routes/usuarios.js
const router = require('express').Router()
const bcrypt = require('bcryptjs')
const db     = require('../config/db')
const { verificarToken, soloAdmin } = require('../middleware/auth')

// Todos los endpoints requieren token
router.use(verificarToken)

// GET /api/usuarios — listar (admin ve todos, encargado solo se ve a sí mismo)
router.get('/', async (req, res) => {
  try {
    let query, params
    if (req.usuario.rol === 'administrador') {
      // Admin ve todos los usuarios con sus laboratorios asignados
      query = `
        SELECT u.id, u.nombre, u.correo, u.rol, u.activo, u.ultimo_acceso,
              GROUP_CONCAT(l.nombre SEPARATOR ', ') AS laboratorios
        FROM usuarios u
        LEFT JOIN usuario_laboratorio ul ON u.id = ul.usuario_id
        LEFT JOIN laboratorios l ON ul.laboratorio_id = l.id
        GROUP BY u.id
        ORDER BY u.nombre`
      params = []
    } else {
      query  = 'SELECT id, nombre, correo, rol, activo, ultimo_acceso FROM usuarios WHERE id = ?'
      params = [req.usuario.id]
    }
    const [rows] = await db.execute(query, params)
    res.json(rows)
  } catch (err) {
    console.error(err)
    res.status(500).json({ error: 'Error al obtener usuarios' })
  }
})

// POST /api/usuarios — crear (solo admin)
router.post('/', soloAdmin, async (req, res) => {
  const { nombre, correo, password, rol, laboratorios } = req.body
  if (!nombre || !correo || !password || !rol)
    return res.status(400).json({ error: 'Faltan campos obligatorios' })

  try {
    const hash = await bcrypt.hash(password, 10)
    const [result] = await db.execute(
      'INSERT INTO usuarios (nombre, correo, password_hash, rol) VALUES (?, ?, ?, ?)',
      [nombre, correo, hash, rol]
    )
    const userId = result.insertId

    // Asignar laboratorios si viene el arreglo
    if (laboratorios && laboratorios.length > 0) {
      const values = laboratorios.map(labId => [userId, labId])
      await db.query('INSERT INTO usuario_laboratorio (usuario_id, laboratorio_id) VALUES ?', [values])
    }

    res.status(201).json({ id: userId, mensaje: 'Usuario creado correctamente' })
  } catch (err) {
    if (err.code === 'ER_DUP_ENTRY')
      return res.status(409).json({ error: 'El correo ya está registrado' })
    console.error(err)
    res.status(500).json({ error: 'Error al crear usuario' })
  }
})

// PUT /api/usuarios/:id — editar (admin edita cualquiera, encargado solo su perfil)
router.put('/:id', async (req, res) => {
  const id = parseInt(req.params.id)
  if (req.usuario.rol !== 'administrador' && req.usuario.id !== id)
    return res.status(403).json({ error: 'Sin permiso' })

  const { nombre, correo, password, rol, activo, laboratorios } = req.body
  try {
    if (password) {
      const hash = await bcrypt.hash(password, 10)
      await db.execute(
        'UPDATE usuarios SET nombre=?, correo=?, password_hash=?, rol=?, activo=? WHERE id=?',
        [nombre, correo, hash, rol, activo ?? 1, id]
      )
    } else {
      await db.execute(
        'UPDATE usuarios SET nombre=?, correo=?, rol=?, activo=? WHERE id=?',
        [nombre, correo, rol, activo ?? 1, id]
      )
    }

    // Reasignar laboratorios si se envían (solo admin)
    if (req.usuario.rol === 'administrador' && laboratorios !== undefined) {
      await db.execute('DELETE FROM usuario_laboratorio WHERE usuario_id = ?', [id])
      if (laboratorios.length > 0) {
        const values = laboratorios.map(labId => [id, labId])
        await db.query('INSERT INTO usuario_laboratorio (usuario_id, laboratorio_id) VALUES ?', [values])
      }
    }

    res.json({ mensaje: 'Usuario actualizado' })
  } catch (err) {
    console.error(err)
    res.status(500).json({ error: 'Error al actualizar usuario' })
  }
})

// DELETE /api/usuarios/:id — desactivar (solo admin)
router.delete('/:id', soloAdmin, async (req, res) => {
  try {
    await db.execute('UPDATE usuarios SET activo = 0 WHERE id = ?', [req.params.id])
    res.json({ mensaje: 'Usuario desactivado' })
  } catch (err) {
    res.status(500).json({ error: 'Error al desactivar usuario' })
  }
})

module.exports = router

File: backend/routes/usuarios.py
import bcrypt
from flask import Blueprint, request, jsonify, g
from config import get_connection
from middleware import verificar_token, solo_admin

usuarios_bp = Blueprint('usuarios', __name__)


# GET /api/usuarios
@usuarios_bp.route('/', methods=['GET'])
@verificar_token
def get_usuarios():
    conn   = get_connection()
    cursor = conn.cursor()
    try:
        if g.usuario.get('rol') == 'administrador':
            cursor.execute("""
                SELECT u.id, u.nombre, u.correo, u.rol, u.activo, u.ultimo_acceso,
                       GROUP_CONCAT(l.nombre ORDER BY l.nombre SEPARATOR ', ') AS laboratorios
                FROM usuarios u
                LEFT JOIN usuario_laboratorio ul ON u.id = ul.usuario_id
                LEFT JOIN laboratorios l ON ul.laboratorio_id = l.id
                GROUP BY u.id, u.nombre, u.correo, u.rol, u.activo, u.ultimo_acceso
                ORDER BY u.nombre
            """)
        else:
            cursor.execute(
                'SELECT id, nombre, correo, rol, activo, ultimo_acceso FROM usuarios WHERE id=?',
                (g.usuario['id'],)
            )
        cols = [d[0] for d in cursor.description]
        rows = [dict(zip(cols, row)) for row in cursor.fetchall()]
        return jsonify(rows)
    except Exception as err:
        print(err)
        return jsonify({'error': 'Error al obtener usuarios'}), 500
    finally:
        conn.close()


# POST /api/usuarios
@usuarios_bp.route('/', methods=['POST'])
@verificar_token
@solo_admin
def crear_usuario():
    data         = request.get_json() or {}
    nombre       = data.get('nombre', '').strip()
    correo       = data.get('correo', '').strip().lower()
    password     = data.get('password', '')
    rol          = data.get('rol', '')
    laboratorios = data.get('laboratorios', [])

    if not nombre or not correo or not password or not rol:
        return jsonify({'error': 'Faltan campos obligatorios'}), 400

    password_hash = bcrypt.hashpw(password.encode(), bcrypt.gensalt()).decode()

    conn   = get_connection()
    cursor = conn.cursor()
    try:
        cursor.execute(
            'INSERT INTO usuarios (nombre, correo, password_hash, rol) VALUES (?,?,?,?)',
            (nombre, correo, password_hash, rol)
        )
        conn.commit()
        cursor.execute('SELECT LAST_INSERT_ID()')
        user_id = int(cursor.fetchone()[0])

        if laboratorios:
            for lab_id in laboratorios:
                cursor.execute(
                    'INSERT INTO usuario_laboratorio (usuario_id, laboratorio_id) VALUES (?,?)',
                    (user_id, lab_id)
                )
            conn.commit()

        return jsonify({'id': user_id, 'mensaje': 'Usuario creado correctamente'}), 201
    except Exception as err:
        err_str = str(err)
        if 'Duplicate' in err_str or 'duplicate' in err_str or '1062' in err_str:
            return jsonify({'error': 'El correo ya esta registrado'}), 409
        print(err)
        return jsonify({'error': 'Error al crear usuario'}), 500
    finally:
        conn.close()


# PUT /api/usuarios/<id>
@usuarios_bp.route('/<int:user_id>', methods=['PUT'])
@verificar_token
def actualizar_usuario(user_id):
    if g.usuario.get('rol') != 'administrador' and g.usuario['id'] != user_id:
        return jsonify({'error': 'Sin permiso'}), 403

    data         = request.get_json() or {}
    nombre       = data.get('nombre')
    correo       = data.get('correo')
    password     = data.get('password')
    rol          = data.get('rol')
    activo       = data.get('activo', 1)
    laboratorios = data.get('laboratorios')

    conn   = get_connection()
    cursor = conn.cursor()
    try:
        if password:
            password_hash = bcrypt.hashpw(password.encode(), bcrypt.gensalt()).decode()
            cursor.execute(
                'UPDATE usuarios SET nombre=?, correo=?, password_hash=?, rol=?, activo=? WHERE id=?',
                (nombre, correo, password_hash, rol, activo, user_id)
            )
        else:
            cursor.execute(
                'UPDATE usuarios SET nombre=?, correo=?, rol=?, activo=? WHERE id=?',
                (nombre, correo, rol, activo, user_id)
            )
        conn.commit()

        if g.usuario.get('rol') == 'administrador' and laboratorios is not None:
            cursor.execute('DELETE FROM usuario_laboratorio WHERE usuario_id=?', (user_id,))
            for lab_id in laboratorios:
                cursor.execute(
                    'INSERT INTO usuario_laboratorio (usuario_id, laboratorio_id) VALUES (?,?)',
                    (user_id, lab_id)
                )
            conn.commit()

        return jsonify({'mensaje': 'Usuario actualizado'})
    except Exception as err:
        print(err)
        return jsonify({'error': 'Error al actualizar usuario'}), 500
    finally:
        conn.close()


# DELETE /api/usuarios/<id>  — desactiva (no borra)
@usuarios_bp.route('/<int:user_id>', methods=['DELETE'])
@verificar_token
@solo_admin
def desactivar_usuario(user_id):
    conn   = get_connection()
    cursor = conn.cursor()
    try:
        cursor.execute('UPDATE usuarios SET activo=0 WHERE id=?', (user_id,))
        conn.commit()
        return jsonify({'mensaje': 'Usuario desactivado'})
    except Exception:
        return jsonify({'error': 'Error al desactivar usuario'}), 500
    finally:
        conn.close()
